### PR TITLE
EQL: Enforce validation of key compatibility

### PR DESCRIPTION
**Pull request by costin**
Currently the engine allows keys with different types to be specified
for joining.
This is problematic since it suggest there is an implicit normalization
being applied which is not the case and further, might not be what the
user expects.
Instead of letting this silent behavior in, the verifier now validates
that at least the types are compatible.
Do note that this requires a follow-up for text vs strings.

### CODE REVIEW

**Comment by costin** [comment]
Notice this PR will break a few tests in the correctness tests which tries to compare a `source_address` (keyword) to an ip field. Which doesn't really work since there are no guarantees the `source_address` is correctly formed (in fact, it isn't - some fields contain `-` and such).

**Review by matriv — COMMENTED** [review]
LGTM.

Does it make sense to have an enhancement in the future and allow casting?

**Comment by costin** [comment]
> Does it make sense to have an enhancement in the future and allow casting?

Yes, I'd like to first wrap this one up and follow-up with an issue.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Verifier.java [hunk 1]
```diff
@@ -236,4 +244,35 @@ public Verifier(Metrics metrics) {
...
+                        doCheckKeyTypes(join, localFailures, currentKey, until.keys().get(keyIndex));
+                    }
+                }
+            }
+        }
+    }
+
+    private static void doCheckKeyTypes(Join join, Set<Failure> localFailures, NamedExpression expectedKey, NamedExpression currentKey) {
+        if (DataTypes.areCompatible(expectedKey.dataType(), currentKey.dataType()) == false) {
+            localFailures.add(fail(currentKey, "{} key [{}] type [{}] is incompatible with key [{}] type [{}]",
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Verifier.java:271]: I would have chosen a slightly different error message, something that would made clearer that the incompatible types are to blame: `key [X] type [Y] has an incompatible type with key [A] type [B]`.

**Review by astefan — APPROVED** [review]
LGTM. Left one comment.

**Comment by matriv** [comment]
@costin fixed the mapping (`keyword` -> `ip`) for the relevant fields, created a new snapshot and tested with the code of this PR and the tests succeed. Please check: #66238

**Review by matriv — COMMENTED** [review]
LGTM

**Review by matriv — APPROVED** [review]

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Verifier.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Verifier.java:271]: What's unclear about the current message?
`Sequence key[X] type[Y] is incompatible with key [A] type [B]` vs 
`key[X] type[Y] has an incompatible type with key [A] type [B]`.

The messages are the same length however I find the former more informative.

**Comment by costin** [comment]
@elasticmachine run elasticsearch-ci/eql-corectness

**Comment by costin** [comment]
@elasticmachine update branch

**Pull request merged**